sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala

```diff
@@ -98,6 +99,11 @@ object RewritePredicateSubquery extends Rule[LogicalPlan] with PredicateHelper {
           val (newCond, inputPlan) = rewriteExistentialExpr(Seq(predicate), p)
           Project(p.output, Filter(newCond.get, inputPlan))
       }
+
+      rewrotePlan transform {
+        case j @ Join(left, right, _, _) if !j.duplicateResolved =>
+          j.copy(right = DedupQueryAttributesInPlans.dedupRight(left, right))
```

Comment:
Can you explain when this becomes a problem? LeftSemi/LeftAnti joins do output attributes from the right hand side of the join. Is the condition messed up?

Reply:
`duplicateResolved` in logical `Join` doesn't consider if LeftSemi/LeftAnti or not.

But as LeftSemi/LeftAnti only do output from one side (`left.output` in `logical.Join`), should we revise `duplicateResolved` for `logical.Join`? Then we don't need to dedup here.

Reply:
@viirya Just for my understanding, Can you please put an example query that depicts this problem ?

Reply:
One example:

    Seq(1 -> "a").toDF("i", "j").write.parquet(path.getCanonicalPath)
    sql(s"CREATE TABLE t1 USING parquet LOCATION '${path.toURI}'")

    val sqlText =
      """
        |SELECT * FROM t1
        |WHERE
        |NOT EXISTS (SELECT * FROM t1)
      """.stripMargin
      val ds = sql(sqlText)
      println(ds.queryExecution.analyzed)
      println(ds.queryExecution.optimizedPlan)


    Project [i#236, j#237]
    +- Filter NOT exists#235 []
       :  +- Project [i#236, j#237]
       :     +- SubqueryAlias t1
       :        +- Relation[i#236,j#237] parquet
       +- SubqueryAlias t1
          +- Relation[i#236,j#237] parquet

    'Join LeftAnti
    :- Relation[i#236,j#237] parquet
    +- Relation[i#236,j#237] parquet

Reply:
I think it is better to fix `duplicateResolved`.

We should add a project when outputs are duplicate. Could you check this for a correlated subquery?

Reply:
Sure. Thanks for suggestion.

Reply:
@viirya Thanks. We should be de-duping the subquery output in case of correlated subquery. Thats why i was wondering about how we are ending up with duplicate attributes. 
@hvanhovell I had a question. Do we anticipate to have future rewrites to convert left-semi/left-anti joins to other types join such as inner join ? Can this be a problem then ?

Reply:
@hvanhovell Because predicate subqueries are rewritten into left semi/anti joins which don't have duplicate outputs. I think you mean correlated scalar subqueries which are rewritten into left outer joins, is it right?

Reply:
@hvanhovell We may not be able to revise `duplicateResolved`.

Even LeftSemi do output from only left side of the join, we still need `duplicateResolved` as false if there are duplicate attributes between left and right sides.

Otherwise, if there is a condition, the condition will be pushdown to left side of the join, because all attribute references in the condition is belonging to one side. It changes the join results.